File: 1026-maximum-difference-between-node-and-ancestor/1026-maximum-difference-between-node-and-ancestor.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int maxAncestorDiff(TreeNode* root) {
        int ans = 0;
        rec(root, ans);
        return ans;
    }
    
    // minval, maxval
    pair<int, int> rec(TreeNode* root, int &ans) {
        int minval = root->val;
        int maxval = root->val;
        if (root->left) {
            auto p = rec(root->left, ans);
            minval = min(minval, p.first);
            maxval = max(maxval, p.second);
        }
        if (root->right) {
            auto p = rec(root->right, ans);
            minval = min(minval, p.first);
            maxval = max(maxval, p.second);
        }
        ans = max(ans, max(abs(root->val - minval), abs(root->val - maxval)));
        return {minval, maxval};
    }
};

File: 1026-maximum-difference-between-node-and-ancestor/1026-maximum-difference-between-node-and-ancestor.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxAncestorDiff(self, root: Optional[TreeNode]) -> int:
        ans = 0
        
        def dfs(node):
            nonlocal ans
            lo, hi = inf, -inf
            if node.left:
                x, y = dfs(node.left)
                lo = min(lo, x)
                hi = max(hi, y)
            if node.right:
                x, y = dfs(node.right)
                lo = min(lo, x)
                hi = max(hi, y)
            ans = max(ans, node.val - lo, hi - node.val)
            lo = min(lo, node.val)
            hi = max(hi, node.val)
            return lo, hi
        
        dfs(root)
        return ans